gcloud-java-pubsub/src/main/java/com/google/cloud/pubsub/MessageConsumerImpl.java

```diff
@@ -0,0 +1,263 @@
+          return timer;
+        }
+
+        @Override
+        public void close(ScheduledExecutorService instance) {
+          instance.shutdown();
+        }
+      };
+
+  private final PubSubOptions pubsubOptions;
+  private final PubSubRpc pubsubRpc;
+  private final PubSub pubsub;
+  private final AckDeadlineRenewer deadlineRenewer;
+  private final String subscription;
+  private final MessageProcessor messageProcessor;
+  private final ScheduledExecutorService timer;
+  private final ExecutorFactory<ExecutorService> executorFactory;
+  private final ExecutorService executor;
+  private final AtomicInteger queuedCallbacks;
+  private final int maxQueuedCallbacks;
+  private final Object futureLock = new Object();
+  private final Runnable consumerRunnable;
+  private boolean closed;
+  private Future<?> scheduledFuture;
+  private PullFuture pullerFuture;
+
+  /**
+   * Default executor factory for the message processor executor. By default a single-threaded
+   * executor is used.
+   */
+  static class DefaultExecutorFactory implements ExecutorFactory<ExecutorService> {
+
+    private final ExecutorService executor = Executors.newSingleThreadExecutor();
+
+    @Override
+    public ExecutorService get() {
+      return executor;
+    }
+
+    @Override
+    public void release(ExecutorService executor) {
+      executor.shutdownNow();
+    }
+  }
+
+  class ConsumerRunnable implements Runnable {
+
+    @Override
+    public void run() {
+      if (closed) {
+        return;
+      }
+      pullerFuture = pubsubRpc.pull(createPullRequest());
+      pullerFuture.addCallback(new PullCallback() {
+        @Override
+        public void success(PullResponse response) {
+          List<com.google.pubsub.v1.ReceivedMessage> messages = response.getReceivedMessagesList();
+          queuedCallbacks.addAndGet(messages.size());
+          for (com.google.pubsub.v1.ReceivedMessage message : messages) {
+            deadlineRenewer.add(subscription, message.getAckId());
+            ReceivedMessage receivedMessage = ReceivedMessage.fromPb(pubsub, subscription, message);
+            executor.execute(ackingRunnable(receivedMessage));
+          }
+          nextPull();
+        }
+
+        @Override
+        public void failure(Throwable error) {
+          if (!(error instanceof CancellationException)) {
+            nextPull();
+          }
+        }
+      });
+    }
+  }
+
+  private MessageConsumerImpl(Builder builder) {
+    this.pubsubOptions = builder.pubsubOptions;
+    this.subscription = builder.subscription;
+    this.messageProcessor = builder.messageProcessor;
+    this.pubsubRpc = pubsubOptions.rpc();
+    this.pubsub = pubsubOptions.service();
+    this.deadlineRenewer = builder.deadlineRenewer;
+    this.queuedCallbacks = new AtomicInteger();
+    this.timer = SharedResourceHolder.get(TIMER);
+    this.executorFactory = firstNonNull(builder.executorFactory, new DefaultExecutorFactory());
```

Comment:
Nit, a small waist creating an instance of DefaultExecutorFactory even when not necessary (could probably be a constant or just use if statement). Fine as is though.